Rewrite PY3-specific code that uses argspec functions from inspect module

terminalmage opened the pull request:
*This pull request was inspired by https://github.com/saltstack/salt/pull/56027.*

### What does this PR do?

[`inspect.getargspec()`](https://docs.python.org/3/library/inspect.html#inspect.getargspec) and [`inspect.getfullargspec()`](https://docs.python.org/3/library/inspect.html#inspect.getfullargspec) are deprecated in Python 3 in favor of [`inspect.Signature`](https://docs.python.org/3/library/inspect.html#inspect.Signature).

To ensure consistent behavior moving forward with Python 3, this PR does the following:

1. Modifies `salt.utils.args` such that the PY3-specific code to build the namedtuple now uses [`inspect.signature()`](https://docs.python.org/3/library/inspect.html#inspect.signature).

2. Updates all calls to [`inspect.getargspec()`](https://docs.python.org/3/library/inspect.html#inspect.getargspec) and [`inspect.getfullargspec()`](https://docs.python.org/3/library/inspect.html#inspect.getfullargspec) elsewhere in the code to ensure that they use the proper PY3-specific code when running under PY3.

### What issues does this PR fix or reference?

Refs: https://github.com/saltstack/salt/pull/56027

### Merge requirements satisfied?

Docs are unnecessary, as this doesn't change how one uses Salt.

Additionally, new tests are not needed because there are already unit tests that confirm proper operation of `salt.utils.args.get_function_argspec()`, and these tests pass with the new code.

### Commits signed with GPG?
No

DmitryKuzmenko commented:
@terminalmage thank you for PR! It looks some failures are related.

terminalmage committed:
Remove "self" param when is_class_method=True

Also expand test cases.

terminalmage commented:
@DmitryKuzmenko I pushed a commit to fix the failing test case. I also checked and realized the testing for this was not as good as I first thought, so I expanded the test cases for get_function_argspec.

DmitryKuzmenko commented:
Thank you! Let's check it in the morning.

s0undt3ch commented on salt/utils/args.py:
```diff
@@ -23,6 +23,9 @@
 from salt.ext import six
 from salt.ext.six.moves import map, zip
 
+if six.PY3:
```
Master is PY3.
So just import it.

terminalmage replied:
Sorry, master what?

It's being imported because it's only used in for PY3 within `get_function_argspec()`. Maybe it makes more sense to do a late import within `get_function_argspec()`.

s0undt3ch replied:
The master branch is no longer a Py2 branch, though _it can_ support Py2.
So, this code will, in fact, always run under Py3, that's why I said not to gate it behind `six.PY3`

terminalmage replied:
OK, I was confused and thought you were referring somehow to `salt-master`. I've updated the import.

s0undt3ch commented:
On a side note, what's the gain of using `__utils__` instead of just importing the module and calling the function?
It's not like the function uses `__context__`...
Not blocking the merge because of this, I'm genuinely interested in the why.

terminalmage commented:
> On a side note, what's the gain of using `__utils__` instead of just importing the module and calling the function?
> It's not like the function uses `__context__`...
> Not blocking the merge because of this, I'm genuinely interested in the why.

Because this allows for people to deploy their own versions of the utils modules, to do things like deploying bugfixes to those modules via `saltutil.sync_utils`.

s0undt3ch commented:
> Because this allows for people to deploy their own versions of the utils modules, to do things like deploying bugfixes to those modules via saltutil.sync_utils.
Meh, ok.

DmitryKuzmenko reviewed (dismissed):
I see nothing that could be related to the `integration.modules.test_cmdmod.CMDModuleTest` so I'm approving this and re-starting the builds.

terminalmage committed:
Rewrite PY3-specific code that uses argspec functions from inspect module

`inspect.getargspec()` and `inspect.getfullargspec()` are deprecated in
Python 3, so this commit does the following:

1. Modifies `salt.utils.args` such that the PY3-specific code to build
   the namedtuple now uses `inspect.signature()`.

2. Updates all calls to `inpect.getargspec()` to use
   `salt.utils.args.get_function_argspec()`, to ensure that they use the
   proper PY3-specific code when running under PY3.

dwoz commented:
re-run full all
